common/util/tree_operations.h

```diff
@@ -0,0 +1,926 @@
+  (node.Children().push_back(std::forward<AdoptedNodeN>(node_n)), ...);
+}
+
+// This node takes/moves subtrees from another node (concatenates).
+// There need not be any relationship between this node and the other.
+//
+// Requires `push_back()` and `clear()` method in the children container.
+template <class T,  //
+          std::enable_if_t<TreeNodeTraits<T>::available &&
+                           !std::is_const_v<T>>* = nullptr>
+void AdoptSubtreesFrom(T& node, T* other) {
+  using tree_operations_internal::TryReserve;
+  auto& src_children = other->Children();
+  TryReserve(node.Children(),
+             std::size(node.Children()) + std::size(src_children));
+  for (auto& src_child : src_children) {
+    node.Children().push_back(std::move(src_child));
+  }
+  other->Children().clear();
+}
+
+// Recursively transform a SrcTree to DstTree.
+// The resulting tree is always StructureEqual() to the original tree.
+//
+// Usage:
+//   auto tree_other = orig_tree.Transform<OtherType>(
+//       [](const SrcTree& node) { return ... });
+//
+// Requires `push_back()` method in the children container.
+template <
+    class DstTree, class SrcTree, class SrcNodeToDstValueFunc,
+    class DstValue_ =
+        std::invoke_result_t<SrcNodeToDstValueFunc, const SrcTree&>,  //
+    std::enable_if_t<TreeNodeTraits<DstTree>::available &&
+                     TreeNodeTraits<SrcTree>::available &&
+                     std::is_constructible_v<DstTree, DstValue_>>* = nullptr>
+DstTree Transform(const SrcTree& src_node, const SrcNodeToDstValueFunc& f) {
+  using tree_operations_internal::TryReserve;
+  // Using invoke() to allow passing SrcTree's method pointers as `f`
+  DstTree dst_node(std::invoke(f, src_node));
+  TryReserve(dst_node.Children(), std::size(src_node.Children()));
+  for (const auto& child : src_node.Children()) {
+    AdoptSubtree(dst_node, Transform<DstTree>(child, f));
+  }
+  return dst_node;
+}
+
+// If this node has exactly one child, replace this node with that child
+// and return true, otherwise, do nothing and return false.
+template <class T,  //
+          std::enable_if_t<TreeNodeTraits<T>::available &&
+                           !std::is_const_v<T>>* = nullptr>
+bool HoistOnlyChild(T& node) {
+  if (std::size(node.Children()) != 1) return false;
+  // Can't do this directly, as assignment to node destroys its child
+  // (`only`) before it is moved.
+  auto only = std::move(*node.Children().begin());
+  node = std::move(only);
+  return true;
+}
+
+// Combines the Nth and (N+1) sibling using a custom function 'joiner' on the
+// nodes' values, and the Nth sibling will adopt N+1's children.
+// The 'joiner' function does: *left = f(*left, *right);
+// The (N+1) sibling will be erased in the process, and every sibling
+// thereafter will be shifted back one position (same inefficiency as shifting
+// vector contents).  This invalidates all iterators after position N,
```

Comment:
This applies for vector-like containers, with contiguous storage.

Reply:
New comment:
```
// The (N+1) sibling will be erased in the process, and every sibling
// thereafter will be shifted back one position. Depending on a children
// container type this can invalidate all interators after position N and
// interators to Nth node children. This applies e.g. for vector-like
// containers, with contiguous storage.
```